MAINTAINERS.adoc

```diff
@@ -0,0 +1,129 @@
+
+=== Notes about the Dependabot config
+
+Dependabot uses different https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates#versioning-strategy[policies] for updating a package's dependencies, depending on whether it's looking at an application or a library.  The documentation is confusing and self-inconsistent on this subject.  But it appears that when Dependabot checks a package, and it finds that some dependency has a newer version available, then:footnote:[It's worth noting that this description is inconsistent with the Dependabot docs linked above.  First, the docs say that "widen" is the default behavior for libraries, but empirically for Rust https://github.com/dependabot/dependabot-core/blob/f2167589e8210d53ac64e784a77c6aa4a14e31d6/cargo/lib/dependabot/cargo/update_checker/requirements_updater.rb#L45-L50[that's not even supported, and "increase-if-necessary" is the default].  Second, the docs largely don't say anything about what happens to Cargo.toml vs. Cargo.lock.  It's not clear why we've seen with applications that Dependabot sometimes updates Cargo.lock and not Cargo.toml.  The docs are self-inconsistent in other ways, too: it's not clear why they use "npm" and "pip" as examples of apps, or why Cargo is an example of a library, when you can ship apps or libraries with either.  The Cargo subsystem clearly supports both (as you'd expect).  The supported policies also don't make sense: the docs claim that Cargo only supports "auto" and "lockfile-only", but "auto" includes "increase", so how could that not be supported?  Also, Dependabot supports 19 backends, but only 7 are present in the table of which modes are supported.]
+
+* For applications (policy "increase"), it updates Cargo.lock to specify the newer version.  It does not always update Cargo.toml.
+* For libraries (policy https://github.com/dependabot/dependabot-core/blob/f2167589e8210d53ac64e784a77c6aa4a14e31d6/cargo/lib/dependabot/cargo/update_checker/requirements_updater.rb#L45-L50[seems to be "increase-if-necessary" for Rust]), it updates Cargo.toml only when the semver constraint in Cargo.toml does not already allow the newer package version.
+
+For applications (such as Oxide's internal Omicron project), we point Dependabot at the root of the https://doc.rust-lang.org/cargo/reference/workspaces.html[Cargo workspace] (the root of the repository).  This causes Dependabot to find Cargo.lock, treat the package as an application, and update Cargo.lock and the Cargo.toml files for all packages in the workspace.
+
+We do not do this in Dropshot.  We _do_ <<_checking_in_cargo_lock_file,check in a Cargo.lock file>>, but we point Dependabot at the individual packages in the Dropshot workspace.  Dependabot does not find the Cargo.lock file and so treats Dropshot as a library, only updating Cargo.toml when _necessary_ to accommodate a new dependency.
```

Comment:
Earlier we made the assertion the following regarding Lockfiles:

> Since the CI tests always run against a particular commit, this allows us to know exactly which dependencies were used for a particular CI run. It also communicates to consumers what we've tested.

If dependabot can update Cargo.toml files without updating Cargo.lock files, hasn't this assertion been broken? If dependabot updates a rev, but Cargo.lock is left out of date, then won't CI runs be inconsistent?

If we need to add the caveat that "Cargo.lock shows what dependencies are used, unless dependabot changed something, in which case it might be out-of-date", IMO the value of keeping a lockfile around drops significantly.

Reply:
Yeah, that's a good point.  Better options here are probably:

- don't check in Cargo.lock
- point dependabot at the whole workspace and treat it like an application
- point dependabot at the whole workspace and try to configure it to treat it like a library.  This seems best to me, but I'm not super optimistic based on the quality of the docs and the time I spent reverse engineering this last time.

Reply:
It turns out that the docs are actually correct:

> The property '#/updates/0/versioning-strategy' value "increase-if-necessary" did not match one of the following values: lockfile-only, auto

From looking at the source a few weeks ago, the "auto" option decides whether it's an app or a library based on the presence of Cargo.lock (Cargo.lock present => application).  So if we use "auto", dropshot will be treated like an application, which means dependabot will be somewhat aggressive about updating dependencies.  If we chose "lockfile-only", I believe it would bump Cargo.lock as _compatible_ updates to dependencies were published, but never update Cargo.toml (which largely defeats the point).

I think I'd like to try letting dependabot treat dropshot like an application.  The downside is that if we have a dep version like 1.2.3, and 1.3 comes out, it will update Cargo.toml to specify 1.3, even though we might be fine with 1.2.3 _and_ 1.3, so we're imposing additional constraints on our consumers.  I'm hopeful that we can avoid this problem by avoiding such specific dependencies in Cargo.toml unless we really need them.  That is: if we leave our dep at 1.2, and 1.3 is compatible with that, dependabot won't update Cargo.toml.  (I'd expect it to update Cargo.lock, and that's fine and probably useful.)